### PR TITLE
Update api.class.php

### PR DESCRIPTION
**Pull request by 4phun**
Hi,

With reference to our conversation earlier, here's my humble contribution to the XML API. Please check it and let me know what do you think.

Added new function ad_hoc_playlist.

### CODE REVIEW

**Comment by wagnered** [comment]
Not pointing fingers, but I've seen several PR's that hadn't had php-cs-fixer
run before commiting changes and pushing to remote.

I've done it too.   Can't say how many times I had to revert local commits, run
php-cs-fixer and recommit.  The positive side is that only one commit is then needed.

**Comment by 4phun** [comment]
@wagnered True that! Thank you for the heads up.

Pull request have been updated, tests are passing now.

**Comment by lachlan-00** [comment]
There's a little bit of overlap in this that i think is partially solved by advanced_search, stats and playlist methods (when you make a smart playlist.)

object_count is also probably the best place to check activity. user_activity in develop is a lot different compared to stable and it was also not previously mandatory.

Let me have a play with this one but i think we should also call it playlist_generate

File: lib/class/api.class.php [hunk 1]
```diff
@@ -2357,4 +2357,107 @@ public static function democratic($input)
...
+     *
+     * @param array $input
+     * $input = array(mode            = (string)  "least_recent_played"|"most_recent_played"|"random"   // optional, default = "least_recent_played"
+     *                filter          = (string)  $filter                                               // optional, LIKE matched to song title
+     *                album_id        = (integer) $album_id                                             // optional
+     *                artist_id       = (integer) $artist_id                                            // optional
+     *                flagged_only    = (integer) 0|1                                                   // optional, default = 0
+     *                format          = (string)  "song"|"index"|"id"                                   // optional, default = "song"
+     */
+    public static function ad_hoc_playlist($input)
```
Comment (lachlan-00) [inline lib/class/api.class.php]: public static function playlist_generate

**Comment by lachlan-00** [comment]
Actually stats does almost all of it but i think we can tweak this to include some advanced searching as well.
image

File: lib/class/api.class.php [hunk 2]
```diff
@@ -2357,4 +2357,107 @@ public static function democratic($input)
...
+        
+        // process parameters
+
+        $select       = array("song.id AS song_id");
+        $from         = array("song");
+        $where        = array();
+        $order        = array();
+        $bound_values = array();
+
+        if (in_array($input[mode], array("least_recent_played", "most_recent_played"), true)) {
```
Comment (lachlan-00) [inline lib/class/api.class.php]: $input[mode] should be ($input['mode']

The input array is like a dictionary

**Comment by 4phun** [comment]
As we agreed earlier, I only meant this to be a conversation stater.

stat has "recent", but not its opposite "least_recent" a.k.a. "oldest". Not sure what would be the best name for it.

**Comment by 4phun** [comment]
Cool. As we agreed earlier, I only meant this to be a conversation stater.

> Actually stats does almost all of it but i think we can tweak this to include some advanced searching as well.

stat has "recent", but not its opposite "least_recent" a.k.a. "oldest". Not sure what would be the best name for it. If you think it's better to add that to stat, that works for me just fine.

File: lib/class/api.class.php (hunk 2 above)
Comment (4phun) [reply lib/class/api.class.php]: Ah, hard to switch my brain off of perl... But of course you're right.

**Comment by lachlan-00** [comment]
like most things it's close but not terribly close. i'll do an edit on this pull and extend out stats as well. that way you get contributor status in the repo! ;)

I think it's good to keep this as sort of an alias method as i'll be using advanced search and stats to generate the xml.

Give a day or so to work on it and we'll put it in asap.

**Comment by lachlan-00** [comment]
I like to think if it's not 'recent' it's 'forgotten'. i'll ping this pull again when i'm done.

**Comment by 4phun** [comment]
Great, thank you very much.

**Comment by lachlan-00** [comment]
I'm going to try and do this one tonight to finish off the api changes.

**Comment by 4phun** [comment]
Great news, looking forward to see it in Ampache!

**Comment by lachlan-00** [comment]
just working in edge on this now.

i've set it to use searches instead of sql and i just need to test all the options but it's working.

**Comment by lachlan-00** [comment]
I've ruined this tree but will merge this now from https://github.com/ampache/ampache/pull/2183

in your local git do this to restore it after i've merge the new pull
```
git pull
git reset --hard upstream/edge
git push --force origin edge
```

**Comment by 4phun** [comment]
Cool! Thank you!

**Pull request closed without merging**